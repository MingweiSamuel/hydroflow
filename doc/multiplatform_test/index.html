<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="`multiplatform_test`"><title>multiplatform_test - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Bold-a2c9cd1067f8b328.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-a23c9d150ad85d8b.css" id="mainThemeStyle"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="multiplatform_test" data-themes="" data-resource-suffix="" data-rustdoc-version="1.72.0-nightly (330727467 2023-06-29)" data-channel="nightly" data-search-js="search-91b58767224854cb.js" data-settings-js="settings-de11bff964e9d4e5.js" data-settings-css="settings-8c76f75bfb6bd192.css" data-theme-light-css="light-0f8c037637f9eb3e.css" data-theme-dark-css="dark-1097f8e92a01e3cf.css" data-theme-ayu-css="ayu-614652228113ac93.css" ><script src="../static.files/storage-59fd9b8ccb335783.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-0795b7d26be81095.js"></script><noscript><link rel="stylesheet" media="(prefers-color-scheme:light)" href="../static.files/light-0f8c037637f9eb3e.css"><link rel="stylesheet" media="(prefers-color-scheme:dark)" href="../static.files/dark-1097f8e92a01e3cf.css"><link rel="stylesheet" href="../static.files/noscript-13285aec31fa243e.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="logo-container" href="../multiplatform_test/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2></h2></nav><nav class="sidebar"><a class="logo-container" href="../multiplatform_test/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2 class="location"><a href="#">Crate multiplatform_test</a></h2><div class="sidebar-elems"><ul class="block"><li class="version">Version 0.0.0</li><li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#attributes">Attribute Macros</a></li></ul></section></div></nav><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">multiplatform_test</a><button id="copy-path" title="Copy item path to clipboard"><img src="../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="srclink" href="../src/multiplatform_test/lib.rs.html#1-240">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><h2 id="multiplatform_test"><a href="#multiplatform_test"><code>multiplatform_test</code></a></h2>
<p>Provides a proc-macro that expands to testing on platforms relevant to
Hydroflow. By default, expands to testing on the host (using the normal
<code>#[test]</code> attribute) and wasm (using <code>#[wasm_bindgen_test]</code>).</p>
<p>For example, the test</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>multiplatform_test::multiplatform_test;

<span class="attr">#[multiplatform_test]
</span><span class="kw">fn </span>my_test() {
  <span class="comment">// ...
</span>}</code></pre></div>
<p>Expands to</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="attr">#[test]
#[wasm_bindgen_test::wasm_bindgen_test]
</span><span class="kw">fn </span>my_test() {
  <span class="comment">// ...
</span>}</code></pre></div>
<h3 id="installation"><a href="#installation">Installation</a></h3><div class="example-wrap"><pre class="language-toml"><code>[dependencies]
multiplatform_test = * # replace with version.
</code></pre></div>
<p>If you’re using <code>wasm</code> naturally you will need to add the <a href="https://crates.io/crates/wasm-bindgen-test-macro/"><code>wasm_bindgen_test</code></a>
crate as a dependency.</p>
<h3 id="usage"><a href="#usage">Usage</a></h3><h4 id="specifying-platforms"><a href="#specifying-platforms">Specifying platforms</a></h4>
<p>There are many platforms which can be specified:</p>
<ul>
<li><code>test</code> - Adds a standard <a href="https://doc.rust-lang.org/reference/attributes/testing.html#the-test-attribute"><code>#[test]</code> attribute</a>.</li>
<li><code>tokio</code> - Adds a <a href="https://docs.rs/tokio/latest/tokio/attr.test.html"><code>#[tokio::test]</code> attribute</a>.</li>
<li><code>async_std</code> - Adds an <a href="https://docs.rs/async-std/latest/async_std/attr.test.html"><code>#[async_std::test]</code> attribute</a>.</li>
<li><code>hydroflow</code> - Adds a <a href="https://docs.rs/hydroflow/latest/hydroflow/attr.test.html"><code>#[hydroflow::test]</code> attribute</a>.</li>
<li><code>wasm</code> - Adds a <a href="https://docs.rs/wasm-bindgen-test/0.3.36/wasm_bindgen_test/attr.wasm_bindgen_test.html"><code>#[wasm_bindgen_test::wasm_bindgen_test]</code> attribute</a>.</li>
<li><code>env_logging</code> - Registers <a href="https://docs.rs/env_logger/latest/env_logger/"><code>env_logger</code></a> for <a href="https://docs.rs/log/latest/log/"><code>log</code>ging</a>.</li>
<li><code>env_tracing</code> - Registers a <a href="https://docs.rs/tracing-subscriber/latest/tracing_subscriber/index.html#reexport.FmtSubscriber"><code>FmtSubscriber</code></a> with an <a href="https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/struct.EnvFilter.html"><code>EnvFilter</code></a> for <a href="https://docs.rs/tracing/latest/tracing/"><code>tracing</code></a>.</li>
</ul>
<p>You can test on a subset of platforms by passing in the platforms in parens:</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>multiplatform_test::multiplatform_test;

<span class="attr">#[multiplatform_test(test)]  </span><span class="comment">// Only test on the standard `#[test]` platform, but enables logging
</span><span class="kw">fn </span>my_test() {
  <span class="comment">// ...
</span>}</code></pre></div>
<p>expands to</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>multiplatform_test::multiplatform_test;

<span class="attr">#[test]
</span><span class="kw">fn </span>my_test() {
  <span class="comment">// ...
</span>}</code></pre></div>
</div></details><h2 id="attributes" class="small-section-header"><a href="#attributes">Attribute Macros</a></h2><ul class="item-table"><li><div class="item-name"><a class="attr" href="attr.multiplatform_test.html" title="attr multiplatform_test::multiplatform_test">multiplatform_test</a></div><div class="desc docblock-short">See the <a href="index.html" title="mod multiplatform_test">crate</a> docs for usage information.</div></li></ul></section></div></main></body></html>