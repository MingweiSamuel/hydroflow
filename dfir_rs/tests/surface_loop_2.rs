// use dfir_rs::util::iter_batches_stream;
use dfir_rs::{assert_graphvis_snapshots, dfir_syntax};
use multiplatform_test::multiplatform_test;

#[multiplatform_test(test, wasm, env_tracing)]
pub fn test_flo_repeat_n_nested() {
    let mut df = dfir_syntax! {
        usrs1 = source_iter(["alice", "bob"]);
        loop {
            usrs2 = usrs1 -> batch()
                -> inspect(|x| println!("{}: {:?}", line!(), x))
                -> flatten()
                -> enumerate();
            loop {
                usrs3 = usrs2 -> repeat_n(3)
                    -> inspect(|x| println!("{}: {:?}", line!(), x))
                    -> flatten()
                    -> enumerate();
                loop {
                    usrs3 -> repeat_n(3)
                        -> inspect(|x| println!("x {:?}", x))
                        -> null();
                        // assert_eq([
                        //     vec!["alice", "bob", "alice", "bob", "alice", "bob"],
                        //     vec!["alice", "bob", "alice", "bob", "alice", "bob"],
                        //     vec!["alice", "bob", "alice", "bob", "alice", "bob"],
                        // ]);
                }
            }
        }
    };
    assert_graphvis_snapshots!(df);
    df.run_available();
}
