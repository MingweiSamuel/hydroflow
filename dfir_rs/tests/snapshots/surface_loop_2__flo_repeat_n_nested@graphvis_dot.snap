---
source: dfir_rs/tests/surface_loop_2.rs
expression: "df.meta_graph().unwrap().to_dot(& Default :: default())"
---
digraph {
    node [fontname="Monaco,Menlo,Consolas,&quot;Droid Sans Mono&quot;,Inconsolata,&quot;Courier New&quot;,monospace", style=filled];
    edge [fontname="Monaco,Menlo,Consolas,&quot;Droid Sans Mono&quot;,Inconsolata,&quot;Courier New&quot;,monospace"];
    n1v1 [label="(n1v1) None source_iter([\"alice\", \"bob\"])", shape=invhouse, fillcolor="#88aaff"]
    n2v1 [label="(n2v1) Some(GraphLoopId(1v1)) batch()", shape=invhouse, fillcolor="#88aaff"]
    n3v1 [label="(n3v1) Some(GraphLoopId(1v1)) inspect(|x| println!(\"{}: {:?}\", line!(), x))", shape=invhouse, fillcolor="#88aaff"]
    n4v1 [label="(n4v1) Some(GraphLoopId(1v1)) flatten()", shape=invhouse, fillcolor="#88aaff"]
    n5v1 [label="(n5v1) Some(GraphLoopId(1v1)) enumerate()", shape=invhouse, fillcolor="#88aaff"]
    n6v1 [label="(n6v1) Some(GraphLoopId(2v1)) repeat_n(3)", shape=invhouse, fillcolor="#88aaff"]
    n7v1 [label="(n7v1) Some(GraphLoopId(2v1)) inspect(|x| println!(\"{}: {:?}\", line!(), x))", shape=invhouse, fillcolor="#88aaff"]
    n8v1 [label="(n8v1) Some(GraphLoopId(2v1)) flatten()", shape=invhouse, fillcolor="#88aaff"]
    n9v1 [label="(n9v1) Some(GraphLoopId(2v1)) enumerate()", shape=invhouse, fillcolor="#88aaff"]
    n10v1 [label="(n10v1) Some(GraphLoopId(3v1)) repeat_n(3)", shape=invhouse, fillcolor="#88aaff"]
    n11v1 [label="(n11v1) Some(GraphLoopId(3v1)) inspect(|x| println!(\"x {:?}\", x))", shape=invhouse, fillcolor="#88aaff"]
    n12v1 [label="(n12v1) Some(GraphLoopId(3v1)) null()", shape=house, fillcolor="#ffff88"]
    n13v1 [label="(n13v1) None handoff", shape=parallelogram, fillcolor="#ddddff"]
    n14v1 [label="(n14v1) None handoff", shape=parallelogram, fillcolor="#ddddff"]
    n15v1 [label="(n15v1) None handoff", shape=parallelogram, fillcolor="#ddddff"]
    n4v1 -> n5v1
    n3v1 -> n4v1
    n2v1 -> n3v1
    n1v1 -> n13v1
    n8v1 -> n9v1
    n7v1 -> n8v1
    n6v1 -> n7v1
    n5v1 -> n14v1
    n11v1 -> n12v1
    n10v1 -> n11v1
    n9v1 -> n15v1
    n13v1 -> n2v1
    n14v1 -> n6v1 [color=red]
    n15v1 -> n10v1 [color=red]
    subgraph "cluster n1v1" {
        fillcolor="#dddddd"
        style=filled
        label = "sg_1v1\nstratum 0"
        n1v1
        subgraph "cluster_sg_1v1_var_usrs1" {
            label="var usrs1"
            n1v1
        }
    }
    subgraph "cluster n2v1" {
        fillcolor="#dddddd"
        style=filled
        label = "sg_2v1\nstratum 0"
        n2v1
        n3v1
        n4v1
        n5v1
        subgraph "cluster_sg_2v1_var_usrs2" {
            label="var usrs2"
            n2v1
            n3v1
            n4v1
            n5v1
        }
    }
    subgraph "cluster n3v1" {
        fillcolor="#dddddd"
        style=filled
        label = "sg_3v1\nstratum 0"
        n6v1
        n7v1
        n8v1
        n9v1
        subgraph "cluster_sg_3v1_var_usrs3" {
            label="var usrs3"
            n6v1
            n7v1
            n8v1
            n9v1
        }
    }
    subgraph "cluster n4v1" {
        fillcolor="#dddddd"
        style=filled
        label = "sg_4v1\nstratum 0"
        n10v1
        n11v1
        n12v1
    }
}
